File: Trees/diameter_of_binary_tree.py
"""
https://leetcode.com/problems/diameter-of-binary-tree/description/
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution: #Time complexity O(n), Space complexity O(n)
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            nonlocal res

            if not root:
                return 0

            leftEdge = dfs(root.left)
            rightEdge = dfs(root.right)

            res = max(res, leftEdge + rightEdge)

            return max(leftEdge, rightEdge) + 1


        dfs(root)
        return res


class MySolution: #Same complexity as above
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node):
            nonlocal res

            if not node:
                return -1

            left = dfs(node.left)
            right = dfs(node.right)

            res = max(res, 2 + left + right)

            return 1 + max(left, right)

        dfs(root)
        return res
        

File: Trees/diameter_of_binary_tree.js
/*
https://leetcode.com/problems/diameter-of-binary-tree/description/
*/


/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     * Time complexity O(n), Space complexity O(n)
     */
    diameterOfBinaryTree(root) {
        let res = 0

        function dfs(node){
            if(!node){
                return -1
            }

            let left = 1 + dfs(node.left)
            let right = 1 + dfs(node.right)

            res = Math.max(res, left + right)
            return Math.max(left, right)
        }

        dfs(root)

        return res
    }
}
